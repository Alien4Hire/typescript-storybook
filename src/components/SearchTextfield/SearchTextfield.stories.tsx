import { ComponentStory, ComponentMeta } from '@storybook/react';
import SearchTextfield from './SearchTextfield';

export default {
  title: 'Component/SearchTextfield',
  component: SearchTextfield,
  argTypes: { onChange: { action: 'text changed' } },
} as ComponentMeta<typeof SearchTextfield>;

const baseArgs = {
  placeholder: 'Search for a Template',
  onChange: () => {},
  onRequestSearch: () => {},
};

const Template: ComponentStory<typeof SearchTextfield> = (args) => (
  <SearchTextfield {...args} />
);

export const Primary = Template.bind({});
Primary.args = { ...baseArgs };
