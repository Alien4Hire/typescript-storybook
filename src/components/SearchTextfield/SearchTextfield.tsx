import { ChangeEventHandler, KeyboardEvent, useState } from 'react';
import { IconButton, InputBase, Paper } from '@mui/material';
import { ReactComponent as Search } from '../../icons/search.svg';
import { theme } from '../../theme';

export type MlSearchTextfieldProps = {
  placeholder: string;
  onChange: (text: string) => void;
  onRequestSearch: (text: string) => void;
};

const styles = {
  container: {
    display: 'flex',
    alignItems: 'center',
    width: '400px',
    height: '34px',
    borderRadius: 0,
    border: `1px solid ${theme.palette.gray.dark}`,
    boxShadow: 'none',
  },
  icon: {
    fill: theme.palette.gray.dark,
  },
  iconButton: {
    width: '34px',
    height: '34px',
    p: '5px',
  },
  textField: {
    ml: 1,
    flex: 1,
  },
};

const SearchTextfield = ({
  placeholder = 'Search for a keyword',
  onChange = () => {},
  onRequestSearch = () => {},
}: MlSearchTextfieldProps) => {
  const { container, icon, iconButton, textField } = styles;
  const [text, setText] = useState('');
  const handleInputChange: ChangeEventHandler<HTMLInputElement> = (e) => {
    const { value } = e.target;
    onChange(value);
    setText(value);
  };

  const preventDefaultForEnterKey = (e: KeyboardEvent<HTMLInputElement>) => {
    if (e.key === 'Enter') {
      e.stopPropagation();
      e.preventDefault();
    }
  };

  return (
    <Paper component="form" sx={container}>
      <InputBase
        sx={textField}
        placeholder={placeholder}
        onChange={handleInputChange}
        value={text}
        data-testid="search-input"
        onKeyDown={preventDefaultForEnterKey}
      />
      <IconButton
        sx={iconButton}
        aria-label="search"
        onClick={() => onRequestSearch(text)}
      >
        <Search style={icon} />
      </IconButton>
    </Paper>
  );
};

export default SearchTextfield;
