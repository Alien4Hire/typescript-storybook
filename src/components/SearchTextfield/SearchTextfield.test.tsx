import { render, fireEvent, screen } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import '../../setupTests';
import renderer from 'react-test-renderer';
import SearchTextfield from './SearchTextfield';

const baseProps = {
  placeholder: 'Search for a keyword',
  onChange: jest.fn(),
  onRequestSearch: jest.fn(),
};

describe('<SearchTextfield />', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <SearchTextfield
          onChange={() => {}}
          onRequestSearch={() => {}}
          placeholder="Search for a Template"
        />
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders without crashing', () => {
    render(<SearchTextfield {...baseProps} />);
  });

  it('should call onChange fn', () => {
    render(<SearchTextfield {...baseProps} />);
    const searchField = screen.getByPlaceholderText(
      'Search for a keyword'
    ) as HTMLInputElement;

    expect(baseProps.onChange).not.toBeCalled();

    fireEvent.change(searchField, { target: { value: 'keyword' } });
    expect(searchField.value).toBe('keyword');
  });

  it('should call onRequestSearch fn', () => {
    render(<SearchTextfield {...baseProps} />);
    const searchButton = screen.getByRole('button');

    expect(baseProps.onRequestSearch).not.toBeCalled();

    fireEvent.click(searchButton);
    expect(baseProps.onRequestSearch).toBeCalledTimes(1);
  });

  it('should not call onChange fn when Enter key pressed', () => {
    render(<SearchTextfield {...baseProps} />);
    const searchField = screen.getByPlaceholderText(
      'Search for a keyword'
    ) as HTMLInputElement;

    expect(baseProps.onChange).not.toBeCalled();

    fireEvent.change(searchField, { target: { key: 'Enter' } });

    expect(baseProps.onRequestSearch).toBeCalledTimes(0);
  });
});
