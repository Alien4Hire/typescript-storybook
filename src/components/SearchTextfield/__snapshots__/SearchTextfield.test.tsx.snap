// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SearchTextfield /> renders correctly 1`] = `
.c0 {
  background-color: #fff;
  color: rgba(0,0,0,0.87);
  -webkit-transition: box-shadow 300ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: box-shadow 300ms cubic-bezier(0.4,0,0.2,1) 0ms;
  border-radius: 4px;
  box-shadow: 0px 2px 1px -1px rgba(0,0,0,0.2),0px 1px 1px 0px rgba(0,0,0,0.14),0px 1px 3px 0px rgba(0,0,0,0.12);
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: 400px;
  height: 34px;
  border-radius: 0;
  border: 1px solid #686868;
  box-shadow: none;
}

.c4 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  position: relative;
  box-sizing: border-box;
  -webkit-tap-highlight-color: transparent;
  background-color: transparent;
  outline: 0;
  border: 0;
  margin: 0;
  border-radius: 0;
  padding: 0;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  vertical-align: middle;
  -moz-appearance: none;
  -webkit-appearance: none;
  -webkit-text-decoration: none;
  text-decoration: none;
  color: inherit;
}

.c4::-moz-focus-inner {
  border-style: none;
}

.c4.Mui-disabled {
  pointer-events: none;
  cursor: default;
}

.c5 {
  text-align: center;
  -webkit-flex: 0 0 auto;
  -ms-flex: 0 0 auto;
  flex: 0 0 auto;
  font-size: 1.5rem;
  padding: 8px;
  border-radius: 50%;
  overflow: visible;
  color: rgba(0,0,0,0.54);
  -webkit-transition: background-color 150ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: background-color 150ms cubic-bezier(0.4,0,0.2,1) 0ms;
  width: 34px;
  height: 34px;
  padding: 5px;
}

.c5:hover {
  background-color: rgba(0,0,0,0.04);
}

.c5.Mui-disabled {
  background-color: transparent;
  color: rgba(0,0,0,0.26);
}

.c1 {
  font-family: "Roboto","Helvetica","Arial",sans-serif;
  font-weight: 400;
  font-size: 1rem;
  line-height: 1.4375em;
  -webkit-letter-spacing: 0.00938em;
  -moz-letter-spacing: 0.00938em;
  -ms-letter-spacing: 0.00938em;
  letter-spacing: 0.00938em;
  color: rgba(0,0,0,0.87);
  box-sizing: border-box;
  position: relative;
  cursor: text;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin-left: 8px;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1.Mui-disabled {
  color: rgba(0,0,0,0.38);
  cursor: default;
}

.c3 {
  font: inherit;
  -webkit-letter-spacing: inherit;
  -moz-letter-spacing: inherit;
  -ms-letter-spacing: inherit;
  letter-spacing: inherit;
  color: currentColor;
  padding: 4px 0 5px;
  border: 0;
  box-sizing: content-box;
  background: none;
  height: 1.4375em;
  margin: 0;
  -webkit-tap-highlight-color: transparent;
  display: block;
  min-width: 0;
  width: 100%;
  -webkit-animation-name: mui-auto-fill-cancel;
  animation-name: mui-auto-fill-cancel;
  -webkit-animation-duration: 10ms;
  animation-duration: 10ms;
}

.c3::-webkit-input-placeholder {
  color: currentColor;
  opacity: 0.42;
  -webkit-transition: opacity 200ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: opacity 200ms cubic-bezier(0.4,0,0.2,1) 0ms;
}

.c3::-moz-placeholder {
  color: currentColor;
  opacity: 0.42;
  -webkit-transition: opacity 200ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: opacity 200ms cubic-bezier(0.4,0,0.2,1) 0ms;
}

.c3:-ms-input-placeholder {
  color: currentColor;
  opacity: 0.42;
  -webkit-transition: opacity 200ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: opacity 200ms cubic-bezier(0.4,0,0.2,1) 0ms;
}

.c3::-ms-input-placeholder {
  color: currentColor;
  opacity: 0.42;
  -webkit-transition: opacity 200ms cubic-bezier(0.4,0,0.2,1) 0ms;
  transition: opacity 200ms cubic-bezier(0.4,0,0.2,1) 0ms;
}

.c3:focus {
  outline: 0;
}

.c3:invalid {
  box-shadow: none;
}

.c3::-webkit-search-decoration {
  -webkit-appearance: none;
}

label[data-shrink=false] + .MuiInputBase-formControl .c2::-webkit-input-placeholder {
  opacity: 0 !important;
}

label[data-shrink=false] + .MuiInputBase-formControl .c2::-moz-placeholder {
  opacity: 0 !important;
}

label[data-shrink=false] + .MuiInputBase-formControl .c2:-ms-input-placeholder {
  opacity: 0 !important;
}

label[data-shrink=false] + .MuiInputBase-formControl .c2::-ms-input-placeholder {
  opacity: 0 !important;
}

label[data-shrink=false] + .MuiInputBase-formControl .c2:focus::-webkit-input-placeholder {
  opacity: 0.42;
}

label[data-shrink=false] + .MuiInputBase-formControl .c2:focus::-moz-placeholder {
  opacity: 0.42;
}

label[data-shrink=false] + .MuiInputBase-formControl .c2:focus:-ms-input-placeholder {
  opacity: 0.42;
}

label[data-shrink=false] + .MuiInputBase-formControl .c2:focus::-ms-input-placeholder {
  opacity: 0.42;
}

.c3.Mui-disabled {
  opacity: 1;
  -webkit-text-fill-color: rgba(0,0,0,0.38);
}

.c3:-webkit-autofill {
  -webkit-animation-duration: 5000s;
  animation-duration: 5000s;
  -webkit-animation-name: mui-auto-fill;
  animation-name: mui-auto-fill;
}

@media print {
  .c4 {
    color-adjust: exact;
  }
}

@media (hover:none) {
  .c5:hover {
    background-color: transparent;
  }
}

<form
  className="c0 MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1"
>
  <div
    className="c1 MuiInputBase-root MuiInputBase-colorPrimary"
    data-testid="search-input"
    onClick={[Function]}
  >
    <input
      className="c2 c3 MuiInputBase-input"
      onAnimationStart={[Function]}
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      placeholder="Search for a Template"
      type="text"
      value=""
    />
  </div>
  <button
    aria-label="search"
    className="c4 MuiButtonBase-root c5 MuiIconButton-root MuiIconButton-sizeMedium"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onContextMenu={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <svg
      style={
        Object {
          "fill": "#686868",
        }
      }
    >
      search.svg
    </svg>
  </button>
</form>
`;
